<a href="https://colab.research.google.com/github/poet5th/2016_GGL_Tasks/blob/master/20251114.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

언팩킹

In [1]:
coordinates = (10, 20)
x, y = coordinates
print(f'x좌표: {x}, y좌표:{y}')

x좌표: 10, y좌표:20


In [4]:
coordinates = (123, 312, 2321)
x, y, z = coordinates
print(f'x좌표: {x}, y좌표: {y}, z좌표: {z}')

x좌표: 123, y좌표: 312, z좌표: 2321


In [23]:
coordinates = (232, 3213, 12321, 1212)
x, y, z, w  = coordinates
print(f'x좌표 : {x}, y좌표 : {y}, z좌표 : {z}, w좌표 : {w}')

x좌표 : 232, y좌표 : 3213, z좌표 : 12321, w좌표 : 1212


In [26]:
coordinates = (232, 3213, 12321, 1212)
coordinates = x, y, z, w
print(f'x좌표 : {x}, y좌표 : {y}, z좌표 : {z}')

x좌표 : 232, y좌표 : 3213, z좌표 : 12321


In [29]:
rgb = [255, 128,0]
red, green, blue = rgb
print(F'빨강:{red}, 초록:{green}, 파랑:{blue}')

빨강:255, 초록:128, 파랑:0


In [30]:
word = 'ABC'
first, second, third = word
print(f'첫번째 문자:{first}, 두번째 문자:{second}, 세번째 문자:{third}')

첫번째 문자:A, 두번째 문자:B, 세번째 문자:C


[코드 2-29] 함수에서 여러 값을 반환하는 언패킹

In [32]:
def get_dataset_stats(values):
    return min(values), max(values), sum(values) / len(values)
## get_dataset_stats는 숫자 리스트를 입력받아 최솟값, 최댓값, 평균값을 계산하여 세 개의 값을 한번에 반환하는 다중 반환 변수이다.
## data 리스트로 get_dataset_stats 함수를 호출하고, 반환된 세 개의 값을 언패킹을 사용해 minimum, maximum, average 변수에
## 각각 할당한다. 마지막으로 f 문자열을 사용하여 이 세 값을 호출한다. 이 때 {average:.2f}는 평균값을 소수점 둘째 자리까지만 표시한다는
## 것을 의미한다.
data = [12, 8, 21, 17, 5, 32, 14]
minimum, maximum, average = get_dataset_stats(data)
print(f'최솟값: {minimum}, 최댓값:{maximum}, 평균: {average:.2f}')

최솟값: 5, 최댓값:32, 평균: 15.57


In [4]:
def get_dataset_stats(values):
    return min(values), max(values), sum(values)/ len(values)
data = [12, 8, 21, 17, 5, 32, 14]
minimum, maximum, average = get_dataset_stats(data)
print(f'최솟값 : {minimum}, 최댓값: {maximum}, 평균: {average:.2f}')

최솟값 : 5, 최댓값: 32, 평균: 15.57


In [8]:
a = 8300000
b = 5448195
c = a - b
print(c)

2851805


In [10]:
RPS = (8300000, 5558195)
x, y = RPS
z = (x-y)/100000
print(z)

27.41805


Mupid가 11랭이 되기 위해선 매일 10만 RP씩 29일동안 먹어야 한다.

In [12]:
def get_dataset_stats(values):
    return min(values), max(values), sum(values)/ len(values)

data = [12, 8, 21, 17, 5, 32, 14]
minimum, maximum, average = get_dataset_stats(data)
print(f'최솟값: {minimum}, 최대값: {maximum}, 평균: {average:.2f}')

최솟값: 5, 최대값: 32, 평균: 15.57


소스 코드 분석제공된 코드는 데이터셋의 최소값, 최대값, 그리고 평균을 계산하고 그 결과를 출력하는 함수 정의 및 실행 부분으로 구성되어 있습니다.

1. 함수 정의Pythondef get_dataset_stats(values):
    return min(values), max(values), sum(values)/ len(values)

def get_dataset_stats(values):: get_dataset_stats라는 이름의 함수를 정의합니다. 이 함수는 하나의 인자(argument)인 values를 입력받는데, 이 values는 일반적으로 숫자들의 리스트(List) 형태의 데이터셋을 의미합니다.

return min(values), max(values), sum(values) / len(values): 함수가 실행된 후, 튜플(Tuple) 형태로 세 가지 값을 반환(return) 합니다.

min(values): 리스트 values 내의 최소값을 계산합니다.

max(values): 리스트 values 내의 최대값을 계산합니다.

sum(values) / len(values): 리스트 values 내의 모든 요소의 합계(sum(values))를 요소의 개수(len(values))로 나누어 평균을 계산합니다.

2. 함수 호출 및 변수 할당Pythondata = [12, 8, 21, 17, 5, 32, 14]
minimum, maximum, average = get_dataset_stats(data)
data = [12, 8, 21, 17, 5, 32, 14]: 계산에 사용할 숫자들로 구성된 리스트를 정의하고 data 변수에 저장합니다.

minimum, maximum, average = get_dataset_stats(data): 정의한 get_dataset_stats 함수를 data 리스트를 인자로 넣어 호출합니다.함수는 (최소값, 최대값, 평균) 형태의 튜플을 반환합니다.Python의 튜플 언패킹(Tuple Unpacking) 기능을 사용하여, 반환된 튜플의 세 요소가 각각 minimum, maximum, average 세 변수에 순서대로 할당됩니다.

변수할당되는 값설명

minimum5 리스트의 최소값 maximum32리스트의 최대값 average(12+8+21+17+5+32+14) / 7 = 15.5714...리스트의 평균3. 결과 출력Pythonprint(f'최솟값: {minimum}, 최대값: {maximum}, 평균: {average:.2f}')
print(...): 최종 결과를 콘솔 화면에 출력하는 함수입니다.f'...': **f-string (포맷팅 문자열 리터럴)**을 사용하여 문자열 내부에 변수 값을 삽입합니다.{minimum}: minimum 변수의 값 (5)이 삽입됩니다.{maximum}: maximum 변수의 값 (32)이 삽입됩니다.{average:.2f}: average 변수의 값 (15.5714...)이 삽입되는데, 여기서 **:.2f**는 소수점 이하 둘째 자리까지 반올림하여 표시하도록 포맷을 지정하는 부분입니다. 따라서 15.57이 출력됩니다.

📝 최종 실행 결과위 코드를 실행하면 최종적으로 다음과 같은 결과가 출력됩니다.최솟값: 5, 최대값: 32, 평균: 15.57

In [14]:
def get_dataset_stats(values):
    return min(values), max(values), sum(values)/ len(values)

data = [12,8,21,17,5,32, 14]
minimum, maximim, average = get_dataset_stats(data)
print(f'최솟값: {minimum}, 최대값 : {maximum}, 평균값: {average:.2f}')

최솟값: 5, 최대값 : 32, 평균값: 15.57


In [15]:
student_scores = [
  ('김철수', 85, 92, 78),
  ('이영희', 92, 88, 95),
  ('박지민', 75, 83, 90)
]

for name, database, python, cloud in student_scores:
    average = (database + python+ cloud) / 3
    print(f'{name}의 평균 점수: {average:.1f}')

김철수의 평균 점수: 85.0
이영희의 평균 점수: 91.7
박지민의 평균 점수: 82.7


[코드 2-30]의 for 문에서는 student_scores 리스트의 각 튜플을 순회하면서, 언패킹을 통해 튜플에 포함된 네 개의 개별 변수에 할당한다. 그리고 세 과목 점수의 평균을 계산하고, f-문자열을 사용하여 학생 이름과 함께 평균 점수를 출력한다.
Python 3에서는 별포(*)를 사용한 확장 언패킹이 추가되었다. 확장 언패킹은 나머지 모든 값을 하나의 리스트로 모을 수 있게 한다. [코드 2-31]은 확장 언패킹을 사용하는 예이다.

In [17]:
monthly_sales = [1200,1350,1420,1500,1300,1500,1620,1700,1800,
                 1850,1900,2000]

first, second, *remaining = monthly_sales
print(f'1월 판매액 : {first}')
print(f'2월 판매액 : {second}')
print(f'나머지 월 판매액 : {remaining}')

1월 판매액 : 1200
2월 판매액 : 1350
나머지 월 판매액 : [1420, 1500, 1300, 1500, 1620, 1700, 1800, 1850, 1900, 2000]


3.예외 처리
대부분의 프로그래밍 언어에서는 코드 실행 중 예기치 않은 오류가 발생할 경우에 대비하는 기능을 제공한다. 이때 오류를 예외(exception)라고 하며, 프로그램이 비정상적으로 종료되지 않도록 대비하는 기능을 예외처리(exception handling)라고 한다. 이를 통해 오류 상황에 적절히 대처할 수 있으며, 프로그램의 안정성을 향상시킬 수 있다. 예외 처리를 하지 않는 경우, 프로그램 실행 중 오류가 발생하면 프로그램이 즉시 종료되기 때문에 사용자 경험 저차, 데이터 손실, 리소스 누수, 디버깅 어려움 등의 문제가 발생할 수 있다.
파이썬은 예외 처리를 위해 try-except 문을 제공하며, 추가적으로 finally 블록을 활용하여 예외 발생 여부와 관계없이 반드시 실행해야 하는 코드를 지정할 수 있게 한다. 우선 예외가 발생하는 상황을 다음 [코드 2-35]를 통해 살펴본다.

In [19]:
number = int(input('정수를 입력하세요:'))
result = 10 / number
print(f'결과: {result}')

정수를 입력하세요:1
결과: 10.0


In [21]:
def add(x):
    return x + 2
add_lambda = lambda x: x + 2
print(add(3))
print(add_lambda(3))

5
5


In [22]:
concat = lambda s1, s2:s1 + ' ' + s2
print(concat('파이썬', '데이터분석'))

파이썬 데이터분석


In [23]:
is_even = lambda x : '짝수' if x %2 == 0 else '홀수'
print(is_even(10))
print(is_even(7))

짝수
홀수


map, filter, reduce

프로그래밍에서 데이터를 효율적으로 변환하고 가공하는 것은 매우 중요한 작업이다.
특히 데이터 분석에서는 리스트나 딕셔너리와 같은 자료구조를 다루면서 연산을 수행해야 하는 경우가 많다. 이러한 경우 반복문을 사용하여 데이터를 처리할 수도 있지만, 파이썬에서는 함수형 프로그래밍 기법을 활용하여 보다 간결하고 효율적인 방식으로 데이터를 변환할 수 있다.
map(), filter(), reduce()는 대표적인 함수형 프로그래밍 기법으로 데이터 변환, 필터링, 누적 연산을 수행하는 역할을 한다. 이 함수들은 반복문을 사용하지 않고도 리스트나 다른 반복 가능한 객체의 요소를 손쉽게 조작할 수 있도록 도와준다. 이를 이해하고 적절하게 활용하면 코드의 가독성이 향상될 뿐만 아니라 연산 속도도 최적화할 수 있다.


map 함수

map 함수는 반복 가능한 객체의 모든 요소에 특정 함수를 적용하여 새로운 값을 생성하는 함수이다. 이 함수는 리스트나 튜플과 같은 자료구조의 요소를 변경해야 할 때 유용하게 사용된다.

다름 [코드 2-45]는 리스트의 각 요소를 다른 값으로 변환하는 기본적인 예이다.


In [24]:
temperature_celsius = [25.6, 27.8, 30.5, 22.3, 28.9, 31.2, 24.7]
celsius_to_fahrenheit = lambda c: (c * 9/5) + 32
temperature_fahrenheit = list(map(celsius_to_fahrenheit, temperature_celsius))

print('섭씨 온도 데이터:', temperature_celsius)
print('화씨 온도 데이터:', temperature_fahrenheit)

섭씨 온도 데이터: [25.6, 27.8, 30.5, 22.3, 28.9, 31.2, 24.7]
화씨 온도 데이터: [78.08, 82.04, 86.9, 72.14, 84.02, 88.16, 76.46]


[코드 2-45]에서는 섭씨 온도 데이터가 저장된 temperature_celsius 리스트를 정의하고 다음으로 섭씨 온도를 인자로 받아서 화씨 온도로 변환하는 람다 한수 celsius_to_fahrenheit를 정의한다. 그 다음 map() 함수는 적용할 함수와 반복 가능한 객체인 리스트를 매개변수로 받아 리스트의 각 요소에 지정된 함수를 적용한 결과를 반환한다. 이때 map() 함수의 결과는 map 객체이므로 list() 함수를 사용하여 리스트로 변환한 뒤 temperature_fahrenheit 변수에 저장한다. 마지막으로, 두 개의 print()문을 사용하여 원래의 섭씨 온도 데이터와 변환된 화씨 온도 데이터를 출력한다.

map 함수는 문자열 데이터를 변환해야 하는 경우에도 사용될 수 있다.



**[코드 2-46] map 함수 사용 예(2)**

In [27]:
words = ['castle sauvage', 'druim ligen', 'svasud faste' ]
uppercase_words = list(map(lambda word: word.upper(), words))
print(uppercase_words)

['CASTLE SAUVAGE', 'DRUIM LIGEN', 'SVASUD FASTE']


In [25]:
words = ['python', 'data', 'analysis']
uppercase_words = list(map(lambda word: word.upper(), words))
print(uppercase_words)

['PYTHON', 'DATA', 'ANALYSIS']


[코드 2-46]에서 lambda word: word.upper()는 문자를 대문자로 변환하는 람다 함수이다. map() 함수를 사용하여 리스트의 각 단어에 람다 함수 lambda word: word.upper()를 적용하고 리스트로 변환한다. 그러면 리스트['PYTHON', 'DATA', 'ANALYSIS']를 최종적으로 출력한다.

딕셔너리 데이터를 다룰 떄도 map()을 활용하면, 특정 값을 변환하는 작업을 간결하게 수행할 수 있다. [코드 2-43]과 유사하게 직원의 급여를 10% 인상하지만, for 문 없이 처리하는 [코드 2-47]을 통해 map()의 기능에 대해 알아본다.


In [30]:
employees = [
    {'name': '김지원', 'salary': 3700000},
    {'name': '박민준', 'salary': 7000000},
    {'name': '이서연', 'salary': 4400000},
    {'name': '정현우', 'salary': 3500000}
]

updated_employees = list(map(lambda emp:{'name': emp['name'],
                              'salary': emp['salary'] * 1.1}, employees))
print('\n급여 인상 후:')
for emp in updated_employees:
    print(f'{emp['name']}: {int(emp['salary']):,}')


급여 인상 후;
김지원: 4,070,000
박민준: 7,700,000
이서연: 4,840,000
정현우: 3,850,000


lambda emp:{'name':, emp['name'], 'salary['salary'] * 1.1}은 직원의 급여를 10% 인상하는 람다 함수이다. map()을 사용하여 기존 딕셔너리의 값을 변경하면서 새로운 리스트를 생성한다. 결과적으로 기존 급여보다 10% 증가한 새로운 급여 리스트가 생성된다.

[] filter 함수

상황에 따라 전체 데이터가 아닌 일부 데이터에만 처리가 요구되는 경우가 있다. 예를 들어, 특정 점수 이상을 받은 학생만 추출하거나, 결측치가 없는 데이터만 남기거나, 특정 조건을 만족하는 고객 정보를 필터링하는 등의 작업을 수행할 수 있다. 이때 필요한 데이터만 추출하는 과정이 필요한데, 파이썬에서는 이를 위해 filter() 함수를 제공한다. filter() 함수는 반복문 없이도 간결하고 효율적인 데이터 필터링을 가능하게 한다.
filter() 함수는 리스트나 튜플과 같은 반복 가능한 객체의 요소 중 특정 조건을 만족하는 요소만 남긴다. True 또는 False를 반환하는 함수(조건 함수)를 첫 번째 인자로 받고, 두 번째 인자로 반복 가능한 객체를 받는다. 그리고 해당 조건 함수가 True를 반환하는 요소만 새로운 컬렉션에 포함시킨다. [코드 2-48]은 간단한 filter() 함수를 통한 필터링이다.